File: packages/ui/src/format.ts
/**
 * Display formatting shared by the web and admin apps.
 *
 * Everything here takes the wire representation the API actually sends — integer paise, an
 * ISO instant with an offset, an IANA timezone from the store — and returns a string. None
 * of it does arithmetic on money beyond dividing by 100 for display: every total the user
 * sees is computed server-side, and a client that recomputes one will eventually disagree
 * with the amount that was charged.
 */

const INR = new Intl.NumberFormat('en-IN', {
  style: 'currency',
  currency: 'INR',
  minimumFractionDigits: 0,
  maximumFractionDigits: 0,
});

const INR_WITH_PAISE = new Intl.NumberFormat('en-IN', {
  style: 'currency',
  currency: 'INR',
  minimumFractionDigits: 2,
  maximumFractionDigits: 2,
});

/**
 * `4900` → `₹49`. Prices in this catalog are whole rupees, so trailing `.00` is noise.
 *
 * Anything with real paise keeps them rather than rounding — a refund of ₹49.50 must not
 * be shown as ₹50, and `en-IN` grouping means ₹1,00,000 rather than ₹100,000.
 */
export function formatMoney(paise: number): string {
  const rupees = paise / 100;
  return Number.isInteger(rupees) ? INR.format(rupees) : INR_WITH_PAISE.format(rupees);
}

/** `4900` → `49`. For inputs and CSV cells, where a currency symbol is in the way. */
export function paiseToRupees(paise: number): number {
  return paise / 100;
}

/** `49.5` → `4950`. Rounds, because floating-point rupees do not divide cleanly. */
export function rupeesToPaise(rupees: number): number {
  return Math.round(rupees * 100);
}

/**
 * `90` → `1h 30m`, `45` → `45m`.
 *
 * Compact because it sits inside chips and table cells where "1 hour 30 minutes" wraps.
 */
export function formatDuration(minutes: number): string {
  if (minutes < 60) return `${minutes}m`;
  const hours = Math.floor(minutes / 60);
  const rest = minutes % 60;
  return rest === 0 ? `${hours}h` : `${hours}h ${rest}m`;
}

function parts(
  iso: string,
  timeZone: string | undefined,
  options: Intl.DateTimeFormatOptions,
): string {
  const date = new Date(iso);
  if (Number.isNaN(date.getTime())) return '—';
  return new Intl.DateTimeFormat('en-IN', { ...options, ...(timeZone ? { timeZone } : {}) })
    .format(date);
}

/**
 * `2026-08-10T20:15:00+05:30` → `8:15 pm`.
 *
 * Pass the store's IANA timezone. The API already emits instants at the store's offset, so
 * omitting it is usually harmless — but only until a customer opens the app abroad, at which
 * point every slot silently shifts. Pass it.
 */
export function formatTime(iso: string, timeZone?: string): string {
  return parts(iso, timeZone, { hour: 'numeric', minute: '2-digit', hour12: true })
    .replace(' AM', ' am')
    .replace(' PM', ' pm');
}

/**
 * `Mon 10 Aug`.
 *
 * `en-IN` renders this as "Mon, 10 Aug". The comma is dropped so `formatDateTime` reads
 * "Mon 10 Aug, 8:15 pm" rather than carrying two commas in one short string.
 */
export function formatDate(iso: string, timeZone?: string): string {
  return parts(iso, timeZone, { weekday: 'short', day: 'numeric', month: 'short' }).replace(
    ',',
    '',
  );
}

/** `Mon 10 Aug, 8:15 pm`. */
export function formatDateTime(iso: string, timeZone?: string): string {
  return `${formatDate(iso, timeZone)}, ${formatTime(iso, timeZone)}`;
}

/** `8:15 – 9:00 pm`. The meridiem appears once when both ends share it. */
export function formatTimeRange(startIso: string, endIso: string, timeZone?: string): string {
  const start = formatTime(startIso, timeZone);
  const end = formatTime(endIso, timeZone);
  const startMeridiem = start.slice(-2);
  return startMeridiem === end.slice(-2)
    ? `${start.slice(0, -3)} – ${end}`
    : `${start} – ${end}`;
}

/**
 * `today` / `tomorrow` / `Mon 10 Aug`, resolved in the store's timezone.
 *
 * Compared as calendar dates rather than by hours elapsed: at 23:30 a booking three hours
 * away is tomorrow, and calling it "today" sends someone to the store on the wrong day.
 */
export function formatRelativeDay(iso: string, timeZone?: string): string {
  const dayKey = (date: Date): string =>
    new Intl.DateTimeFormat('en-CA', {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      ...(timeZone ? { timeZone } : {}),
    }).format(date);

  const target = new Date(iso);
  if (Number.isNaN(target.getTime())) return '—';

  const now = new Date();
  const tomorrow = new Date(now.getTime() + 86_400_000);
  const yesterday = new Date(now.getTime() - 86_400_000);

  const key = dayKey(target);
  if (key === dayKey(now)) return 'Today';
  if (key === dayKey(tomorrow)) return 'Tomorrow';
  if (key === dayKey(yesterday)) return 'Yesterday';
  return formatDate(iso, timeZone);
}

/**
 * Seconds remaining until an instant, floored at zero.
 *
 * Used by the hold countdown. Returns a number rather than a string so the caller decides
 * how to render nothing-left — a "00:00" that lingers reads as a broken timer.
 */
export function secondsUntil(iso: string, now: Date = new Date()): number {
  const target = new Date(iso).getTime();
  if (Number.isNaN(target)) return 0;
  return Math.max(0, Math.floor((target - now.getTime()) / 1000));
}

/** `95` → `1:35`. Always two digits of seconds. */
export function formatCountdown(totalSeconds: number): string {
  const safe = Math.max(0, Math.floor(totalSeconds));
  const minutes = Math.floor(safe / 60);
  const seconds = safe % 60;
  return `${minutes}:${String(seconds).padStart(2, '0')}`;
}

/**
 * `+919404491801` → `+91 94044 91801`.
 *
 * Indian numbers only; anything else is returned untouched rather than grouped wrongly.
 */
export function formatPhone(e164: string): string {
  const match = /^\+91(\d{5})(\d{5})$/.exec(e164);
  return match === null ? e164 : `+91 ${match[1]} ${match[2]}`;
}

/** `RST2K8F4M` → `RST-2K8F4M`. Grouped so it can be read aloud at a counter. */
export function formatBookingCode(publicId: string): string {
  return publicId.includes('-') ? publicId : publicId.replace(/^([A-Z]{3})(.+)$/, '$1-$2');
}

/** `12.5` → `12.5%`, `12` → `12%`. */
export function formatPercent(value: number): string {
  return `${Number.isInteger(value) ? value : value.toFixed(1)}%`;
}

/** `1536` → `1.5 KB`. For the media library. */
export function formatBytes(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

/** `YYYY-MM-DD` for a Date, in the given timezone. The format every date query wants. */
export function toLocalDate(date: Date, timeZone?: string): string {
  return new Intl.DateTimeFormat('en-CA', {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    ...(timeZone ? { timeZone } : {}),
  }).format(date);
}

/** Today in the store's timezone, as `YYYY-MM-DD`. */
export function todayLocal(timeZone?: string): string {
  return toLocalDate(new Date(), timeZone);
}

/** Shifts a `YYYY-MM-DD` by whole days without going through a timezone at all. */
export function addDays(localDate: string, days: number): string {
  const [y, m, d] = localDate.split('-').map(Number);
  // Built in UTC deliberately: a local-time Date would shift across a DST boundary and
  // "tomorrow" would occasionally be the same day again.
  const date = new Date(Date.UTC(y ?? 1970, (m ?? 1) - 1, d ?? 1));
  date.setUTCDate(date.getUTCDate() + days);
  return date.toISOString().slice(0, 10);
}

File: apps/mobile/lib/src/format.dart
import 'package:intl/intl.dart';

/// Display formatting.
///
/// Mirrors `packages/ui/src/format.ts` so the app and the website say the same thing about
/// the same booking. Nothing here does arithmetic on money beyond dividing by 100 to show
/// it: every total is computed server-side, and a client that recomputes one will
/// eventually disagree with what was charged.

final _inr = NumberFormat.currency(locale: 'en_IN', symbol: '₹', decimalDigits: 0);
final _inrPaise = NumberFormat.currency(locale: 'en_IN', symbol: '₹', decimalDigits: 2);

/// The store's UTC offset.
///
/// Dart has no timezone database in the core library, and — the trap this exists for —
/// `DateTime.parse('2026-08-10T20:15:00+05:30')` returns a *UTC* DateTime and throws the
/// offset away. Formatting that directly prints 2:45 pm for a 8:15 pm slot.
///
/// Every instant the API sends already carries the store's offset, so it is captured from
/// the first one parsed and used to shift instants back to store wall-clock for display.
/// Comparisons still use the true instant, which is what `isAfter` and `difference` need.
///
/// A single mutable value is right here because there is exactly one store. It defaults to
/// IST so the very first frame, before any response has landed, is not an hour wrong.
Duration _storeOffset = const Duration(hours: 5, minutes: 30);

Duration get storeOffset => _storeOffset;

/// Records the offset carried by an API instant. Called from model parsing.
void rememberStoreOffset(String iso) {
  final match = RegExp(r'([+-])(\d{2}):?(\d{2})$').firstMatch(iso);
  if (match == null) return;

  final sign = match.group(1) == '-' ? -1 : 1;
  _storeOffset = Duration(
    hours: sign * int.parse(match.group(2)!),
    minutes: sign * int.parse(match.group(3)!),
  );
}

/// An instant shifted into the store's wall-clock, ready for the field-based formatters.
DateTime _inStoreTime(DateTime instant) => instant.toUtc().add(_storeOffset);

/// `4900` → `₹49`. Prices here are whole rupees, so trailing `.00` is noise — but a refund
/// of ₹49.50 keeps its paise rather than being rounded into a different number.
String formatMoney(int paise) {
  final rupees = paise / 100;
  return rupees == rupees.roundToDouble()
      ? _inr.format(rupees)
      : _inrPaise.format(rupees);
}

/// `90` → `1h 30m`. Compact, because it sits inside chips and list rows.
String formatDuration(int minutes) {
  if (minutes < 60) return '${minutes}m';
  final hours = minutes ~/ 60;
  final rest = minutes % 60;
  return rest == 0 ? '${hours}h' : '${hours}h ${rest}m';
}

/// `8:15 pm`, in store time.
///
/// Never `.toLocal()`: that renders in the *device's* timezone, so a customer abroad — or
/// one whose phone clock is simply wrong — sees a time nobody at the counter recognises.
String formatTime(DateTime instant) {
  return DateFormat('h:mm a')
      .format(_inStoreTime(instant))
      .replaceAll('AM', 'am')
      .replaceAll('PM', 'pm');
}

/// `Mon 10 Aug`, in store time.
String formatDate(DateTime instant) =>
    DateFormat('EEE d MMM').format(_inStoreTime(instant));

String formatDateTime(DateTime instant) =>
    '${formatDate(instant)}, ${formatTime(instant)}';

/// `8:15 – 9:00 pm`. The meridiem appears once when both ends share it.
String formatTimeRange(DateTime start, DateTime end) {
  final a = formatTime(start);
  final b = formatTime(end);
  return a.substring(a.length - 2) == b.substring(b.length - 2)
      ? '${a.substring(0, a.length - 3)} – $b'
      : '$a – $b';
}

/// `Today` / `Tomorrow` / `Mon 10 Aug`, resolved in store time.
///
/// Compared as calendar dates rather than by hours elapsed: at 23:30 a booking three hours
/// away is tomorrow, and calling it "today" sends someone in on the wrong day. Both sides
/// are shifted into store time first, so the comparison is between two store-local dates
/// rather than one store date and one device date.
String formatRelativeDay(DateTime instant, {DateTime? now}) {
  final target = _inStoreTime(instant);
  final base = _inStoreTime(now ?? DateTime.now());

  final diff = DateTime(target.year, target.month, target.day)
      .difference(DateTime(base.year, base.month, base.day))
      .inDays;

  return switch (diff) {
    0 => 'Today',
    1 => 'Tomorrow',
    -1 => 'Yesterday',
    _ => formatDate(instant),
  };
}

/// `YYYY-MM-DD` in store time — the format every date query wants.
String formatIsoDate(DateTime date) =>
    DateFormat('yyyy-MM-dd').format(_inStoreTime(date));

/// `95` → `1:35`. Always two digits of seconds.
String formatCountdown(Duration remaining) {
  final total = remaining.isNegative ? 0 : remaining.inSeconds;
  final minutes = total ~/ 60;
  final seconds = total % 60;
  return '$minutes:${seconds.toString().padLeft(2, '0')}';
}

/// `+919404491801` → `+91 94044 91801`. Anything not Indian is returned untouched rather
/// than grouped wrongly.
String formatPhone(String e164) {
  final match = RegExp(r'^\+91(\d{5})(\d{5})$').firstMatch(e164);
  return match == null ? e164 : '+91 ${match.group(1)} ${match.group(2)}';
}

/// `RST2K8F4M` → `RST-2K8F4M`, so it can be read aloud at a counter.
String formatBookingCode(String publicId) {
  if (publicId.contains('-')) return publicId;
  return publicId.replaceFirstMapped(
    RegExp(r'^([A-Z]{3})(.+)$'),
    (m) => '${m.group(1)}-${m.group(2)}',
  );
}

/// Accepts what people type — `9404491801`, `+91 94044 91801`, `094044 91801` — and
/// produces the E.164 the API requires.
String toE164(String input) {
  final digits = input.replaceAll(RegExp(r'\D'), '');
  final national = digits.length > 10 ? digits.substring(digits.length - 10) : digits;
  return '+91$national';
}
